// pangea-node-sdk/dist/commonjs/services/authn/client/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_js_1 = __importDefault(require("../../../services/base.js"));
const session_js_1 = __importDefault(require("./session.js"));
const password_js_1 = __importDefault(require("./password.js"));
const token_js_1 = __importDefault(require("./token.js"));
class Client extends base_js_1.default {
    session;
    password;
    clientToken;
    constructor(token, config) {
        super("authn", token, config);
        this.session = new session_js_1.default(token, config);
        this.password = new password_js_1.default(token, config);
        this.clientToken = new token_js_1.default(token, config);
    }
    /**
     * @summary Get User (client token)
     * @description Retrieve the logged in user's token and information.
     * @operationId authn_post_v2_client_userinfo
     * @param {String} code - Login code returned by the login callback
     * @returns {Promise<PangeaResponse<AuthN.Client.UserinfoResult>>} - A promise
     * representing an async call to the endpoint. Available response fields can be found in our
     * [API Documentation](https://pangea.cloud/docs/api/authn/user#/v2/client/userinfo).
     * @example
     * ```js
     * const response = await authn.client.userinfo(
     *   "pmc_d6chl6qulpn3it34oerwm3cqwsjd6dxw"
     * );
     * ```
     */
    userinfo(code) {
        const data = {
            code: code,
        };
        return this.post("v2/client/userinfo", data);
    }
    /**
     * @summary Get JWT verification keys
     * @description Get JWT verification keys.
     * @operationId authn_post_v2_client_jwks
     * @returns {Promise<PangeaResponse<AuthN.Client.JWKSResult>>} - A promise
     * representing an async call to the endpoint. Available response fields can be found in our
     * [API Documentation](https://pangea.cloud/docs/api/authn/jwt#/v2/client/jwks).
     * @example
     * ```js
     * const response = await authn.client.jwks();
     * ```
     */
    jwks() {
        return this.post("v2/client/jwks", {});
    }
}
exports.default = Client;
