// pangea-node-sdk/dist/esm/response.js
import fs from "node:fs";
import path from "node:path";
import { getBoundary, parse } from "./utils/multipart.js";
const SupportedJSONFields = ["message", "new", "old"];
const decoder = new TextDecoder("utf-8");
/** Pangea response object. */
export class ResponseObject {
    request_id = "InvalidPayloadReceived";
    request_time = "InvalidPayloadReceived";
    response_time = "InvalidPayloadReceived";
    status = "NoStatus";
    result = {};
    accepted_result;
    summary = "InvalidPayloadReceived";
    constructor(body) {
        Object.assign(this, body);
    }
}
export class AttachedFile {
    filename;
    file;
    contentType;
    constructor(filename, file, contentType) {
        this.filename = filename;
        this.file = file;
        this.contentType = contentType;
    }
    save(destFolder, filename) {
        if (!destFolder) {
            destFolder = ".";
        }
        if (!filename) {
            filename = this.filename ? this.filename : "defaultName.txt";
        }
        if (!fs.existsSync(destFolder)) {
            // If it doesn't exist, create it
            fs.mkdirSync(destFolder, { recursive: true });
        }
        const filepath = path.resolve(destFolder, filename);
        fs.writeFileSync(filepath, Buffer.from(this.file));
    }
}
export class PangeaResponse extends ResponseObject {
    gotResponse;
    success;
    attachedFiles = [];
    constructor(response, body) {
        let jsonResp = {};
        let attachedFilesTemp = [];
        const contentType = response.headers.get("content-type");
        if (contentType && contentType.includes("multipart")) {
            const boundary = getBoundary(contentType);
            const parts = parse(Buffer.from(body), boundary);
            parts.forEach((part, index) => {
                if (index == 0) {
                    jsonResp = JSON.parse(part.data.toString("utf-8"));
                }
                else {
                    attachedFilesTemp.push(new AttachedFile(part.filename, part.data, part.type));
                }
            });
        }
        else {
            jsonResp = JSON.parse(decoder.decode(body), parseJSONfields);
        }
        super(jsonResp);
        this.attachedFiles = attachedFilesTemp;
        this.gotResponse = response;
        this.success = this.status === "Success";
        this.result = this.result == null ? {} : this.result;
        if (this.gotResponse.status === 202) {
            this.accepted_result = this.result;
            this.result = {};
        }
    }
    // Return raw Pangea API response body
    body() {
        return this.gotResponse?.body;
    }
    // Return SDK PangeaResponse object as JSON (excluded raw API response body)
    toJSON() {
        return JSON.stringify({ ...this, gotResponse: undefined }, null, "  ");
    }
}
function parseJSONfields(key, value) {
    if (SupportedJSONFields.includes(key) && typeof value === "string") {
        try {
            const obj = JSON.parse(value);
            return obj;
        }
        catch (e) {
            return value;
        }
    }
    else {
        return value;
    }
}
export default PangeaResponse;
