// pangea-node-sdk/dist/services/authn/index.js
import BaseService from "../../services/base.js";
import User from "./user/index.js";
import Flow from "./flow/index.js";
import Client from "./client/index.js";
import Session from "./session.js";
import Agreements from "./agreements/index.js";
/**
 * AuthnService class provides methods for interacting with the AuthN Service
 * @extends BaseService
 */
class AuthNService extends BaseService {
    constructor(token, config) {
        super("authn", token, config);
        Object.defineProperty(this, "user", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "flow", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "session", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "agreements", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.user = new User(token, config);
        this.flow = new Flow(token, config);
        this.client = new Client(token, config);
        this.session = new Session(token, config);
        this.agreements = new Agreements(token, config);
    }
}
export default AuthNService;
