// pangea-node-sdk/dist/services/file_scan.js
import BaseService from "./base.js";
import PangeaConfig from "../config.js";
import { TransferMethod, } from "../types.js";
import { getFileUploadParams } from "../utils/utils.js";
import { PangeaErrors } from "../errors.js";
import PangeaRequest from "../request.js";
export class FileScanService extends BaseService {
    /**
     * Creates a new `FileScanService` with the given Pangea API token and
     * configuration.
     *
     * @param token Pangea API token.
     * @param config Configuration.
     *
     * @example
     * ```js
     * const config = new PangeaConfig({ domain: "pangea_domain" });
     * const client = new FileScanService("pangea_token", config);
     * ```
     *
     * @summary File Scan
     */
    constructor(token, config) {
        super("file-scan", token, config);
    }
    /**
     * @summary Scan
     * @description Scan a file for malicious content.
     * @operationId file_scan_post_v1_scan
     * @param {FileScan.ScanRequest} request
     * @param {string} filepath
     * @param {FileScan.Options} options
     * @returns {Promise} - A promise representing an async call to the check endpoint
     * @example
     * ```js
     * const request = { verbose: true, raw: true, provider: "crowdstrike" };
     * const response = await client.fileScan(request, "./path/to/file.pdf");
     * ```
     */
    fileScan(request, file, // This param is optional. It should be null when using the source_url method
    options = {
        pollResultSync: true,
    }) {
        let fsData = {};
        if (request.transfer_method === TransferMethod.PUT_URL) {
            throw new PangeaErrors.PangeaError(`${request.transfer_method} not supported in this function. Use getUploadURL() instead.`);
        }
        let postFile = undefined;
        let files = undefined;
        if (typeof file === "string") {
            postFile = {
                name: "file",
                file: file,
            };
        }
        else {
            postFile = file;
        }
        if (postFile) {
            files = {
                file: postFile,
            };
        }
        const postOptions = {
            pollResultSync: options.pollResultSync,
            files: files,
        };
        if ((!request.transfer_method ||
            request.transfer_method === TransferMethod.POST_URL) &&
            postFile) {
            fsData = getFileUploadParams(postFile.file);
        }
        const fullRequest = {
            ...fsData,
            ...request,
        };
        return this.post("v1/scan", fullRequest, postOptions);
    }
    // TODO: Docs
    async requestUploadURL(request, options = {}) {
        if (request.transfer_method === TransferMethod.POST_URL &&
            !options.params) {
            throw new PangeaErrors.PangeaError(`If transfer_method is ${TransferMethod.POST_URL} need to set options.params`);
        }
        let fsParams = {};
        if (request.transfer_method === TransferMethod.POST_URL && options.params) {
            fsParams = options.params;
        }
        const fullRequest = {
            ...fsParams,
            ...request,
        };
        return await this.request.requestPresignedURL("v1/scan", fullRequest);
    }
}
export class FileScanUploader {
    serviceName = "FileScanFileUploader";
    request_ = undefined;
    constructor() { }
    get request() {
        if (this.request_) {
            return this.request_;
        }
        this.request_ = new PangeaRequest(this.serviceName, "unusedtoken", new PangeaConfig());
        return this.request_;
    }
    // TODO: Docs
    async uploadFile(url, fileData, options) {
        if (!options.transfer_method ||
            options.transfer_method === TransferMethod.PUT_URL) {
            await this.request.putPresignedURL(url, fileData);
        }
        else if (options.transfer_method === TransferMethod.POST_URL) {
            await this.request.postPresignedURL(url, fileData);
        }
    }
}
