// pangea-node-sdk/dist/commonjs/services/authn/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_js_1 = __importDefault(require("../../services/base.js"));
const index_js_1 = __importDefault(require("./user/index.js"));
const index_js_2 = __importDefault(require("./flow/index.js"));
const index_js_3 = __importDefault(require("./client/index.js"));
const session_js_1 = __importDefault(require("./session.js"));
const index_js_4 = __importDefault(require("./agreements/index.js"));
/**
 * AuthnService class provides methods for interacting with the AuthN Service
 * @extends BaseService
 */
class AuthNService extends base_js_1.default {
    user;
    flow;
    client;
    session;
    agreements;
    /**
     * Creates a new `AuthNService` with the given Pangea API token and
     * configuration.
     *
     * @param token Pangea API token.
     * @param config Configuration.
     *
     * @example
     * ```js
     * const config = new PangeaConfig({ domain: "pangea_domain" });
     * const authn = new AuthNService("pangea_token", config);
     * ```
     *
     * @summary AuthN
     */
    constructor(token, config) {
        super("authn", token, config);
        this.user = new index_js_1.default(token, config);
        this.flow = new index_js_2.default(token, config);
        this.client = new index_js_3.default(token, config);
        this.session = new session_js_1.default(token, config);
        this.agreements = new index_js_4.default(token, config);
    }
}
exports.default = AuthNService;
